modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java

```diff
@@ -310,7 +330,168 @@ void run(ClusterState state) {
+                    Map<String, String> lifecycleMetadata = backingIndex.getCustomData(LIFECYCLE_CUSTOM_INDEX_METADATA_KEY);
+                    if (lifecycleMetadata == null || lifecycleMetadata.containsKey(REPLACEMENT_SOURCE_INDEX) == false) {
+                        // this is a downsampling index that was not added by data stream lifecycle in this data stream
+                        continue;
+                    }
+                    String actualDownsamplingSource = lifecycleMetadata.get(REPLACEMENT_SOURCE_INDEX);
+                    IndexMetadata downsampleSourceIndex = metadata.index(actualDownsamplingSource);
+                    if (downsampleSourceIndex != null) {
+                        affectedIndices.add(downsampleSourceIndex.getIndex());
+                        // delete downsampling source index (that's not part of the data stream anymore) before doing any more
+                        // downsampling
+                        DeleteIndexRequest deleteIndexRequest = new DeleteIndexRequest(backingIndexDownsamplingSource).masterNodeTimeout(
+                            TimeValue.MAX_VALUE
+                        );
+                        transportActionsDeduplicator.executeOnce(
+                            deleteIndexRequest,
+                            new ErrorRecordingActionListener(backingIndexDownsamplingSource, errorStore),
+                            (req, reqListener) -> deleteIndex(
+                                deleteIndexRequest,
+                                "replacement with its downsampled index in the data stream",
+                                reqListener
+                            )
+                        );
+                        continue;
+                    }
+                }
+
+                // this index has matching downsample rounds and is read-only, so let's wait for an in-progress downsampling operation to
+                // succeed or trigger the corresponding one
+                // TODO add support for cancelling a current in-progress operation if another, later, round matches
+                DataStreamLifecycle.Downsampling.Round lastRound = downsamplingRounds.get(downsamplingRounds.size() - 1);
+
+                for (DataStreamLifecycle.Downsampling.Round round : downsamplingRounds) {
+                    // the downsample index name for each round is deterministic
+                    String downsampleIndexName = DownsampleConfig.generateDownsampleIndexName(
+                        DOWNSAMPLED_INDEX_PREFIX,
+                        backingIndex,
+                        round.config().getFixedInterval()
+                    );
+                    IndexMetadata downsampleIndexMeta = metadata.index(downsampleIndexName);
+                    if (downsampleIndexMeta != null) {
+                        IndexMetadata.DownsampleTaskStatus downsampleStatus = INDEX_DOWNSAMPLE_STATUS.get(
```

Comment:
Can this ever be null? Should we protect against that (tossing an error into the error store if it's the case) just to be on the safe side?

Reply:
The setting has a default value so it can't be `null`. (the `UNKNOW` status is the default)